documentation/src/main/asciidoc/performance_guide/performance_guide.asciidoc

```diff
@@ -0,0 +1,241 @@
+* `-XX:+UseConcMarkSweepGC` - Enables usage of the CMS collector.
+* `-XX:+CMSClassUnloadingEnabled` - Allows class unloading when the CMS collector is enabled.
+* `-XX:+UseParNewGC` - Utilize a parallel collector for the young generation. This parameter minimizes pausing by using multiple collection threads in parallel.
+* `-XX:+DisableExplicitGC` - Prevent explicit garbage collections.
+* `-XX:+UseG1GC` - Turn on G1 Garbage Collector.
+
+=== Other settings
+
+There are two additional parameters which are suggested to be used:
+
+* `-server` - Enables server mode for the JVM.
+* `-XX:+ UseLargePages` - Instructs the JVM to allocate memory in Large Pages. These pages must be configured at the OS level for this parameter to function successfully.
+
+=== Example configuration
+
+In most of the cases we suggest using CMS. However when using the latest JVM, G1 might perform slightly better.
+
+.32GB JVM
+----
+-server
+-Xmx32G
+-Xms32G
+-Xmn8G
+-XX:+UseLargePages
+-XX:+UseConcMarkSweepGC
+-XX:+UseParNewGC
+-XX:+DisableExplicitGC
+----
+
+.32GB JVM with G1 Garbage Collector
+----
+-server
+-Xmx32G
+-Xms32G
+-Xmn8G
+-XX:+UseG1GC
+----
+
+== Network configuration
+
+Infinispan uses TCP/IP for sending packets over the network (for both cluster communication when using TCP stack or when communication with Hot Rod clients)
+
+In order to achieve the best results, it is recommended to increase TCP send and receive window size (refer to you OS manual for instructions).
+The recommended values are:
+
+* send window size - 640 KB
+* receive window size - 25 MB
+
+== Number of threads (Server mode only)
+
+Hot Rod Server uses worker threads which are activated by a client's requests.
+It's important to match the number of worker threads to the number of concurrent client requests:
+
+[source,java]
+.Hot Rod Server worker thread pool size
+----
+<hotrod-connector socket-binding="hotrod" cache-container="local" worker-threads="200">
+   <!-- Additional configuration here -->
+</hotrod-connector>
+----
+
+== Cache Store performance
+
+In order to achieve the best performance, please follow the recommendations below when using Cache Stores:
+
+* Use async mode (write-behind) if possible
+* Prevent cache misses by preloading data
+* For JDBC Cache Store:
+** Use indexes on `id` column to prevent table scans
+** Use PRIMARY_KEY on `id` column
+** Configure batch-size, fetch-size, etc
+
+== Hints for program developers
+
+There are also several hints for developers which can be easily applied to the client application and will boost up the performance.
+
+=== Ignore return values
+
+When you're not interested in returning value of the `#put(k, v)` or `#remove(k)` method, use `Flag.IGNORE_RETURN_VALUES` flag as shown below:
+
+[source,java]
+.Flag.IGNORE_RETURN_VALUES
+----
+Cache noPreviousValueCache = cache.getAdvancedCache().withFlags(Flag.IGNORE_RETURN_VALUES);
+noPreviousValueCache.put(k, v);
+----
```

Comment:
Should we mention that this flag can be turned on everywhere using the **unsafe** configuration options ?